midend/interpreter.cpp

```diff
@@ -309,6 +309,81 @@ void SymbolicStruct::dbprint(std::ostream& out) const {
     out << " }";
 }
 
+SymbolicHeaderUnion::SymbolicHeaderUnion(const IR::Type_HeaderUnion* type,
+                               bool uninitialized,
+                               const SymbolicValueFactory* factory) :
+        SymbolicStruct(type, uninitialized, factory),
+        valid(new SymbolicBool(false)) {}
+
+void SymbolicHeaderUnion::setValid(bool v) {
+    if (!v)
+        setAllUnknown();
+    valid = new SymbolicBool(v);
+}
+
+SymbolicValue* SymbolicHeaderUnion::get(const IR::Node* node, cstring field) const {
+    if (valid->isKnown() && !valid->value)
+        return new SymbolicStaticError(node, "Reading field from invalid header union");
+    return SymbolicStruct::get(node, field);
+}
+
+void SymbolicHeaderUnion::setAllUnknown() {
+    SymbolicStruct::setAllUnknown();
+    valid->setAllUnknown();
+}
+
+SymbolicValue* SymbolicHeaderUnion::clone() const {
+    auto result = new SymbolicHeaderUnion(type->to<IR::Type_HeaderUnion>());
+    for (auto f : fieldValue)
+        result->fieldValue[f.first] = f.second->clone();
+    result->valid = valid->clone()->to<SymbolicBool>();
+    return result;
+}
+
+void SymbolicHeaderUnion::assign(const SymbolicValue* other) {
+    if (other->is<SymbolicError>()) return;
+    BUG_CHECK(other->is<SymbolicHeaderUnion>(), "%1%: expected a header union", other);
+    auto hv = other->to<SymbolicHeaderUnion>();
+    for (auto f : hv->fieldValue)
+        fieldValue[f.first]->assign(f.second);
+    valid->assign(hv->valid);
+}
+
+bool SymbolicHeaderUnion::merge(const SymbolicValue* other) {
+    BUG_CHECK(other->is<SymbolicHeaderUnion>(), "%1%: expected a header union", other);
+    auto hv = other->to<SymbolicHeaderUnion>();
```

Comment:
Used checkedTo for all of these checks.

Reply:
`P4::SymbolicValue` has no `checkedTo ` function

Reply:
If it is not too much effort, I would just add it then. Clearly, there is a use case. Alternatively, do the cast with `to` first, then check if `to` is a nullptr. So you do not have to cast twice.

Reply:
Ok, i`ll do this